### DISCUSSION TITLE
Add the ability to set a foreign key on a model to create relationships

### DISCUSSION THREAD

**Discussion by canrozanes** [mswjs/data#297] [Ideas]
It would be great if developers could be a given the option to create one-to-one relationships using foreign keys. This pattern should feel natural for developers who are familiar with relational databases. 

```ts
import { factory, primaryKey, foreignKey } from '@mswjs/data'

const db = factory({
  user: {
    id: primaryKey(String),
    publications: manyOf('post'),
  },
  post: {
    id: primaryKey(String),
    userId: foreignKey("user")
    title: String,
  },
})
```

And then we could get posts belonging to a specific user with the following syntax.

```ts
const user = db.user.findFirst({
   where: {
      id: "abc"
   }
})

const userBlogPosts = db.post.findMany({
   where: {
      user_id: user.id
   }
})
```

Alternatively, the library could implicitly create a `userId` field on a blogPost if blogPost was declared with a `oneOf("user")` relationship. I think this is what [mirage](https://miragejs.com/docs/main-concepts/relationships/) does.

**Comment by kettanaito** [comment]
Hi, @canrozanes. Thanks for raising this. 

Tell me, would the usage of transient + derivative properties work for you?

```js
const db = factory({
  user: {
    id: id(Number),
    // Create a transient "user.city" property.
    // It can be given initial value but it will never
    // be set on the user entities.
    city: transient(oneOf('city')),
    // Derive a public "user.cityId" foreign key
    // from the transient "user.city" one-of relationship.
    cityId: derivative(({ city }) => city.name),
  },
  city: {
    name: id(String),
  },
})

const user = db.user.create({
  id: 1,
  city: db.city.create({ name: 'London' }),
})

user.cityId // "London"
user.city // undefined
```

> This is a non-existing API (yet). 

I'm tackling this precise use case in the Data rewrite and I could use your feedback. 

## Why not use `foreignKey()`?

The API you proposed implies that the key given as the value to the `user_id` property would be identifiable as foreign. But in the context of operations (create/update/etc), those key are taken from the existing _extraneous_ models (like `user`), and will know nothing about `post.userId` having to be foreign. 

To put plainly, it becomes problematic to ensure the contract with this kind of API.

**Reply by canrozanes** [reply]
Based on the examples you provided, `transient` and `derivative` should work. 

Would it work similarly for manyOf relationships? I was thinking something along the lines of:

```ts
const db = factory({
  user: {
    id: id(Number),
    cities: transient(manyOf('city')),
    cityIds: derivative(({ cities }) => cities.map((city)=>city.id)),
  },
  city: {
    id: id(String),
  },
})
```

**Reply by nbibler** [reply]
This seems like a fair approach to me, as well. 👍

**Comment by kettanaito** [comment]
I wanted to give a quick update on this: I'm finalizing the revamp of the Data, and relationship declaration will be finally fixed. I believe in conventional database table design, you get a foreign ID column first, and then define a relationship over its value. This is how all ORM tools I found work. 

I will share more updates once I open the pull request. Need to finalize a few design decisions, battle through TypeScript, add tests, and then we will have it. Thank you for your patience.

**Reply by nbibler** [reply]
Nice! Thank you for all of your effort in that!

**Reply by kettanaito** [reply]
I know this is taking some time but I'm doing my best for it to be worth it. Thanks for the support!
